pkg/daemons/control/tunnel.go

```diff
@@ -148,11 +150,21 @@ func (t *TunnelServer) onChangePod(podName string, pod *v1.Pod) (*v1.Pod, error)
 			for _, ip := range pod.Status.PodIPs {
 				if cidr, err := util.IPStringToIPNet(ip.IP); err == nil {
 					if pod.DeletionTimestamp != nil {
-						logrus.Debugf("Tunnel server egress proxy removing Node %s Pod IP %v", nodeName, cidr)
-						t.cidrs.Remove(*cidr)
+						nets, _ := t.cidrs.ContainingNetworks(net.ParseIP(ip.IP))
+						if entry, ok := nets[0].(*tunnelEntry); ok {
+							if entry.podID == pod.UID {
+								// only remove when the latest owner pod is deleted, otherwise a Complete Pod (job) may cause deleting the entry
+								logrus.Debugf("Tunnel server egress proxy removing Node %s Pod IP %v", nodeName, cidr)
+								t.cidrs.Remove(*cidr)
+							}
+						}
```

Comment:
nit: possible use of `nets[0]` without validating slice length. You're discarding the error and not checking the length before referencing the first element.
Also not checking the ParseIP return but I think you can just use `cidr` here?

Ref: https://github.com/k3s-io/k3s/blob/c86b66d05cef4ed7a9d9dde2e96d376d1ba640ba/pkg/daemons/control/tunnel.go#L202-L206

```suggestion
						if nets, err := t.cidrs.ContainingNetworks(cidr); err == nil && len(nets) > 0 {
							if entry, ok := nets[0].(*tunnelEntry); ok {
								if entry.podID == pod.UID {
									// only remove when the latest owner pod is deleted, otherwise a Complete Pod (job) may cause deleting the entry
									logrus.Debugf("Tunnel server egress proxy removing Node %s Pod IP %v", nodeName, cidr)
									t.cidrs.Remove(*cidr)
								}
							}
						}
```

Reply:
Thanks -  `nets, _ := t.cidrs.ContainingNetworks(cidr.IP)` avoids extra parsing & error check.

As for the `if entry, ok := nets[0].(*tunnelEntry); ok {` I copied it from `dialBackend` assuming no risk of bad data.